web/pdf_attachment_viewer.js

```diff
@@ -49,152 +46,144 @@ var PDFAttachmentViewer = (function PDFAttachmentViewerClosure() {
-      var attachmentsCount = 0;
-
-      if (this.attachments) {
-        var keepRenderedCapability = params.keepRenderedCapability === true;
-        this.reset(keepRenderedCapability);
-      }
-      this.attachments = attachments;
+  /**
+   * @private
+   */
+  _dispatchEvent(attachmentsCount) {
+    this.eventBus.dispatch('attachmentsloaded', {
+      source: this,
+      attachmentsCount,
+    });
 
-      if (!attachments) {
-        this._dispatchEvent(attachmentsCount);
-        return;
+    this._renderedCapability.resolve();
+  }
+
+  /**
+   * @private
+   */
+  _bindPdfLink(button, content, filename) {
+    var blobUrl;
+    button.onclick = () => {
+      if (!blobUrl) {
+        blobUrl = createObjectURL(
+          content, 'application/pdf', PDFJS.disableCreateObjectURL);
+      }
+      var viewerUrl;
+      if (typeof PDFJSDev === 'undefined' || PDFJSDev.test('GENERIC')) {
+        // The current URL is the viewer, let's use it and append the file.
+        viewerUrl = '?file=' + encodeURIComponent(blobUrl + '#' + filename);
+      } else if (PDFJSDev.test('CHROME')) {
+        // In the Chrome extension, the URL is rewritten using the history API
+        // in viewer.js, so an absolute URL must be generated.
+        // eslint-disable-next-line no-undef
+        viewerUrl = chrome.runtime.getURL('/content/web/viewer.html') +
+          '?file=' + encodeURIComponent(blobUrl + '#' + filename);
+      } else {
+        // Let Firefox's content handler catch the URL and display the PDF.
+        // In Firefox PDFJS.disableCreateObjectURL is always false, so
+        // blobUrl is always a blob:-URL and never a data:-URL.
+        viewerUrl = blobUrl + '?' + encodeURIComponent(filename);
       }
+      window.open(viewerUrl);
+      return false;
+    };
+  }
 
-      var names = Object.keys(attachments).sort(function(a, b) {
-        return a.toLowerCase().localeCompare(b.toLowerCase());
-      });
-      attachmentsCount = names.length;
-
-      for (var i = 0; i < attachmentsCount; i++) {
-        var item = attachments[names[i]];
-        var filename = removeNullCharacters(getFilenameFromUrl(item.filename));
-
-        var div = document.createElement('div');
-        div.className = 'attachmentsItem';
-        var button = document.createElement('button');
-        button.textContent = filename;
-        if (/\.pdf$/i.test(filename)) {
-          this._bindPdfLink(button, item.content, filename);
-        } else {
-          this._bindLink(button, item.content, filename);
-        }
+  /**
+   * @private
+   */
+  _bindLink(button, content, filename) {
+    button.onclick = (e) => {
+      this.downloadManager.downloadData(content, filename, '');
+      return false;
+    };
+  }
 
-        div.appendChild(button);
-        this.container.appendChild(div);
-      }
+  /**
+   * @param {PDFAttachmentViewerRenderParameters} params
+   */
+  render(params = Object.create(null)) {
```

Comment:
Since `params` should never be undefined when the method is called, I think that just `render(params = {}) {` would suffice and it's also slightly more readable in my opinion.